### PR TITLE
devicestate: add initial Remodel support

### PR DESCRIPTION
**Pull request by mvo5**
This initial PR adds a new devicestate.Remodel() function that
takes a new model assertion and creates a change that will move
the system from the current model to the new model.

In this version of the remodel support is limited to just adding
new required snaps and to change the kernel track.

### CODE REVIEW

File: overlord/devicestate/devicestate.go [hunk 1]
```diff
@@ -286,3 +291,85 @@ func CanManageRefreshes(st *state.State) bool {
...
+			return nil, err
+		}
+		tss = append(tss, ts)
+	}
+	// adjust snaps
+	for _, snapName := range new.RequiredSnaps() {
+		_, err := snapstate.CurrentInfo(st, snapName)
+		// if the snap is not installed we need to install it now
+		if _, ok := err.(*snap.NotInstalledError); ok {
+			ts, err := snapstateInstall(st, snapName, "", snap.R(0), userID, snapstate.Flags{})
```
Comment (zyga) [inline overlord/devicestate/devicestate.go]: If it is installed it may need to follow another track, is that something we want to add a TODO for?

Comment (mvo5) [reply overlord/devicestate/devicestate.go]: Good point! We have no language yet to support this but we should probably just follow the `snap=track` we already have for kernels etc.

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -78,6 +78,8 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	runner.AddHandler("generate-device-key", m.doGenerateDeviceKey, nil)
 	runner.AddHandler("request-serial", m.doRequestSerial, nil)
 	runner.AddHandler("mark-seeded", m.doMarkSeeded, nil)
+	// FIXME: do we need an undo handler here?
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: I think we need strive to make this always the last task in relevant changes. We cannot undo this operation really if it's change a revision change (there is no revert for assertions). That means tough we probably need a comment here and were we add about this fact that it *must* be last.

File: overlord/devicestate/devicestate.go [hunk 3]
```diff
@@ -286,3 +291,85 @@ func CanManageRefreshes(st *state.State) bool {
...
+	// snap of type "gadget/kernel" is allowed so this needs work
+	if current.Kernel() != new.Kernel() {
+		return nil, fmt.Errorf("cannot remodel to different kernels yet")
+	}
+	if current.Gadget() != new.Gadget() {
+		return nil, fmt.Errorf("cannot remodel to different gadgets yet")
+	}
+	userID := 0
+
+	// adjust tracks
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go]: don't think we need to cover everything in the first iteration but we should have an ongoing TODO list of missing bits, here is a start:

* Need new session/serial if changing store or model
* Check all relevant snaps exist in new store (need to check that even unchanged snaps are accessible)
* Check estimated disk size  delta
* Download everything in a first phase of the change and "pin" cache files (also get assertions), which means
  also dealing with new based and content providers
* Reapply gadget connections as needed

File: overlord/devicestate/devicestate.go [hunk 4]
```diff
@@ -286,3 +291,85 @@ func CanManageRefreshes(st *state.State) bool {
...
+	if current.KernelTrack() != new.KernelTrack() {
+		ts, err := snapstate.Update(st, new.Kernel(), new.KernelTrack(), snap.R(0), userID, snapstate.Flags{})
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go]: is this tested

File: overlord/devicestate/devicestate.go [hunk 5]
```diff
@@ -286,3 +291,85 @@ func CanManageRefreshes(st *state.State) bool {
...
+	// adjust snaps
+	for _, snapName := range new.RequiredSnaps() {
+		_, err := snapstate.CurrentInfo(st, snapName)
+		// if the snap is not installed we need to install it now
+		if _, ok := err.(*snap.NotInstalledError); ok {
+			ts, err := snapstateInstall(st, snapName, "", snap.R(0), userID, snapstate.Flags{})
+			if err != nil {
+				return nil, err
+			}
+			tss = append(tss, ts)
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go]: I think we want to be consistent with how first boot works make these sequential (once we split out downloads first those will not need to be)

File: overlord/devicestate/handlers.go [hunk 6]
```diff
@@ -57,6 +57,23 @@ func (m *DeviceManager) doMarkSeeded(t *state.Task, _ *tomb.Tomb) error {
...
+	var modelass []byte
+	if err := t.Get("new-model", &modelass); err != nil {
+		return err
+	}
+	model, err := asserts.Decode(modelass)
+	if err != nil {
+		return err
+	}
+
+	return assertstate.Add(st, model)
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: this fails we same assertion is added again,  we need to ignore that error to make this idempotent (there are examples about that in other places, let me know if you need help to locate one), we need tests too about running this twice

File: overlord/managers_test.go [hunk 7]
```diff
@@ -2770,3 +2736,140 @@ func (ms *mgrsSuite) TestDisconnectOnUninstallRemovesAutoconnection(c *C) {
...
+	for k, v := range modelExtra {
+		modelV[k] = v
+	}
+	model, err := brandSigning.Sign(asserts.ModelType, modelV, nil, "")
+	c.Assert(err, IsNil)
+
+	return model.(*asserts.Model)
+}
+
+func (ms *mgrsSuite) TestRemodelRequiredSnapsAdded(c *C) {
```
Comment (pedronis) [inline overlord/managers_test.go]: thanks for having tests at this level

File: overlord/devicestate/devicemgr.go [hunk 8]
```diff
@@ -78,6 +78,8 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 	runner.AddHandler("generate-device-key", m.doGenerateDeviceKey, nil)
 	runner.AddHandler("request-serial", m.doRequestSerial, nil)
 	runner.AddHandler("mark-seeded", m.doMarkSeeded, nil)
+	// this *must* always runs last and finalizes a remodel
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: s/runs/run/

File: overlord/devicestate/devicestate.go [hunk 9]
```diff
@@ -286,3 +292,101 @@ func CanManageRefreshes(st *state.State) bool {
...
+// transition is not possible.
+//
+// TODO:
+// - Check estimated disk size delta
+// - Reapply gadget connections as needed
+// - Need new session/serial if changing store or model
+// - Check all relevant snaps exist in new store
+//   (need to check that even unchanged snaps are accessible)
+// - Download everything in a first phase of the change and "pin" cache
+//   files (also get assertions), which means also dealing with new based
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go]: s/based/bases/

File: overlord/devicestate/devicestate.go [hunk 10]
```diff
@@ -286,3 +292,101 @@ func CanManageRefreshes(st *state.State) bool {
...
+			if err != nil {
+				return nil, err
+			}
+			addTss(ts)
+		} else if err != nil {
+			return nil, err
+		}
+	}
+
+	// set the new model assertion
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go]: we want a comment here as well that this must be the last thing done by the change

File: overlord/devicestate/devicestate_test.go [hunk 11]
```diff
@@ -2595,3 +2600,158 @@ func (s *deviceMgrSuite) TestDevicemgrCanStandby(c *C) {
...
+	// waits for first install
+	c.Assert(tss[1].Tasks()[0].WaitTasks(), DeepEquals, tss[0].Tasks())
+
+	c.Assert(tss[2].Tasks()[0].Kind(), Equals, "set-model")
+	c.Assert(tss[2].Tasks()[0].Summary(), Equals, "Set new model assertion")
+	// waits for everything in the change
+	c.Assert(tss[2].Tasks()[0].WaitTasks(), DeepEquals, []*state.Task{tss[0].Tasks()[0], tss[1].Tasks()[0]})
+}
+
+func (s *deviceMgrSuite) TestRemodelSwitchKernelTrack(c *C) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go]: maybe good to also add  a new required snap, to check the chaining as well?

File: overlord/devicestate/handlers.go [hunk 12]
```diff
@@ -57,6 +57,50 @@ func (m *DeviceManager) doMarkSeeded(t *state.Task, _ *tomb.Tomb) error {
...
+	}
+
+	// TODO: look at serial/keyid/sessionmacaroonq
+	device, err := auth.Device(st)
+	if err != nil {
+		return err
+	}
+
+	// set device,model from the new model assertion
+	return setDeviceFromModelAssertion(st, device, model)
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: this bit feels slightly premature,  for now we don't change brand/model

File: overlord/devicestate/handlers.go [hunk 13]
```diff
@@ -57,6 +57,50 @@ func (m *DeviceManager) doMarkSeeded(t *state.Task, _ *tomb.Tomb) error {
...
+	}
+
+	ass, err := asserts.Decode(modelass)
+	if err != nil {
+		return err
+	}
+
+	model, ok := ass.(*asserts.Model)
+	if !ok {
+		return fmt.Errorf("new-model is not a model assertion but: %s", ass.Type().Name)
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: this is an internal error at this point, no? we should fail in Remodel if that was not the case?

File: overlord/devicestate/handlers.go [hunk 14]
```diff
@@ -57,6 +57,50 @@ func (m *DeviceManager) doMarkSeeded(t *state.Task, _ *tomb.Tomb) error {
...
+	}
+
+	err = assertstate.Add(st, ass)
+	if err != nil && !isSameAssertsRevision(err) {
+		return err
+	}
+
+	// TODO: look at serial/keyid/sessionmacaroonq
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: s/sessionmacaroonq/session macaroon/

File: overlord/devicestate/handlers.go (hunk 12 above)
Comment (mvo5) [reply overlord/devicestate/handlers.go]: I can remove it again if you feel it should not (yet) be there.

File: overlord/devicestate/handlers.go (hunk 13 above)
Comment (mvo5) [reply overlord/devicestate/handlers.go]: Yes, I will update the error.

File: overlord/devicestate/handlers.go (hunk 12 above)
Comment (pedronis) [reply overlord/devicestate/handlers.go]: yes, it's premature (we would need to install the new serial that we acquired very early in this case)

**Pull request merged**